<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="use_sim_time" default="true"/>
  <arg name="visualization" default="true"/>
  
  <!-- Parameters for all nodes -->
  <param name="use_sim_time" value="$(arg use_sim_time)"/>
  
  <!-- Path Planner Node -->
  <node pkg="path_planning_pkg" type="path_planner_node" name="path_planner" output="screen">
    <!-- Path planning parameters -->
    <param name="resolution" value="0.2"/>                <!-- Resolution for path planning (meters) -->
    <param name="safety_distance" value="0.5"/>          <!-- Safety distance from obstacles (meters) -->
    <param name="max_planning_time" value="1.0"/>        <!-- Maximum planning time in seconds -->
    <param name="heuristic_weight" value="1.5"/>         <!-- Weight for A* heuristic -->
    <param name="use_3d_connectivity" value="true"/>     <!-- Use full 3D connectivity or just 2D + up/down -->
    
    <!-- Remapping -->
    <remap from="/octomap_binary" to="/octomap_full"/>
    <remap from="/current_state_est" to="/current_state_est"/>
    <remap from="/path_goal" to="/exploration/goal"/>
  </node>
  
  <!-- Trajectory Generator Node -->
  <node pkg="path_planning_pkg" type="trajectory_generator_node" name="trajectory_generator" output="screen">
    <!-- Trajectory parameters -->
    <param name="max_velocity" value="2.0"/>             <!-- m/s -->
    <param name="max_acceleration" value="1.0"/>         <!-- m/s^2 -->
    <param name="max_jerk" value="0.5"/>                 <!-- m/s^3 -->
    <param name="target_dt" value="0.1"/>                <!-- Time step for trajectory points (seconds) -->
    <param name="lookahead_time" value="0.5"/>           <!-- How far ahead to look for waypoints -->
    <param name="min_waypoint_distance" value="0.5"/>    <!-- Minimum distance between waypoints -->
    <param name="yaw_rate_limit" value="0.5"/>           <!-- Maximum yaw rate (rad/s) -->
    
    <!-- Remapping -->
    <remap from="/planned_path" to="/path_planner/path"/>
    <remap from="/current_state_est" to="/current_state_est"/>
    <remap from="/trajectory" to="/desired_trajectory"/>
  </node>
  
  <!-- Exploration Node -->
  <node pkg="exploration_pkg" type="exploration_node" name="exploration" output="screen">
    <!-- Exploration parameters -->
    <param name="exploration_resolution" value="0.5"/>           <!-- meters -->
    <param name="min_frontier_size" value="10"/>                <!-- minimum points in a frontier cluster -->
    <param name="frontier_clustering_distance" value="1.0"/>    <!-- meters -->
    <param name="min_obstacle_distance" value="0.8"/>           <!-- meters -->
    <param name="goal_tolerance" value="0.5"/>                  <!-- meters -->
    <param name="max_objects" value="4"/>                       <!-- maximum number of objects to find -->
    <param name="object_detection_radius" value="1.0"/>         <!-- meters -->
    <param name="return_when_complete" value="true"/>           <!-- return to entrance when done -->
    <param name="world_frame_id" value="world"/>                <!-- coordinate frame -->
    <param name="visualization_enabled" value="$(arg visualization)"/>
    
    <!-- Remapping -->
    <remap from="/octomap_binary" to="/octomap_full"/>
    <remap from="/current_state_est" to="/current_state_est"/>
    <remap from="/semantic_camera/detections" to="/object_detections"/>
    <remap from="/path_goal" to="/exploration/goal"/>
  </node>
  
  <!-- Optional Visualizations -->
  <group if="$(arg visualization)">
    <!-- RViz for visualization -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find path_planning_pkg)/rviz/path_planning.rviz"/>
    
    <!-- Path Visualization -->
    <node pkg="path_planning_pkg" type="path_visualization_node" name="path_visualization">
      <param name="path_color_r" value="0.0"/>
      <param name="path_color_g" value="1.0"/>
      <param name="path_color_b" value="0.0"/>
      <param name="path_alpha" value="1.0"/>
      <param name="trajectory_color_r" value="1.0"/>
      <param name="trajectory_color_g" value="0.0"/>
      <param name="trajectory_color_b" value="1.0"/>
      <param name="trajectory_alpha" value="0.7"/>
    </node>
  </group>
</launch>